src/tainting/Dataflow_tainting.ml

```diff
@@ -1348,36 +1341,70 @@ let check_function_signature env fun_exp args args_taints =
             in
             taints
             |> List.concat_map (fun t ->
-                   let+ dst_taints =
+                   let dst_taints =
                      match t.T.orig with
-                     | Src _ -> Some (Taints.singleton t)
+                     | Src _ -> [ t ]
                      | Arg src_arg ->
                          (* Taint is flowing from one argument to another argument
                           * (or possibly the callee object). Given the formal poly
                           * taint 'src_arg', we compute the actual taint in the
                           * context of this function call. *)
-                         taints_of_sig_arg env fparams fun_exp args args_taints
-                           src_arg
-                         |> Option.map
-                              (Taints.map (fun taint ->
-                                   let tokens =
-                                     List.rev_append t.tokens
-                                       (snd dst_obj.ident :: taint.tokens)
-                                   in
-                                   { taint with tokens }))
+                         first_pass_substitute_arg_with_taints src_arg
+                         |> Common.map (fun taint ->
+                                let tokens =
+                                  List.rev_append t.tokens
+                                    (snd dst_obj.ident :: taint.T.tokens)
+                                in
+                                { taint with tokens })
                    in
-                   if Taints.is_empty dst_taints then []
-                   else [ `UpdateEnv (dst_lval, dst_taints) ])
+                   match dst_taints with
+                   | [] -> []
+                   | _ -> [ `UpdateEnv (dst_lval, Taints.of_list dst_taints) ])
+      in
+      (* This is our second pass for taint substitution, which will run
+          on all of the taints equally. This is because we need to do this
+          on Src taints, even if they weren't the results of substitutions,
+          because they may have dependencies on Arg taints. The alternative
+          is to specifically do the substitution in each case, which is easy
+          to overlook.
+
+          This step could be done on the args taints before we substitute,
```

Comment:
Can't this be done in `process_sig` already even if we don't do it directly on `args_taints` ?

Reply:
We're basically just calling `taints_of_sig_arg` on the `Arg` taints of the preconditions.

Reply:
Yes, but I think it needs to be done uniformly for every case, meaning we would have to call `T.map_preconditions substitute_preconditions` in every branch, which is easy to forget, and also doesn't have a type-level distinction. I personally believe it's better to isolate this step and do it for every single case, so we can't miss it. What do you think?

Reply:
It does get complicated because of the various ways that one needs to tweak the call traces.